{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b080c40e-3418-4b87-9a96-2e1d56d01495",
   "metadata": {},
   "source": [
    "# Model Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0fba5ae0-7d32-444c-bef6-3fa7917a0f45",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model loaded successfully!\n",
      "Model reactions:\n",
      "EX_adn_e\n",
      "EX_but_e\n",
      "EX_chol_e\n",
      "EX_creat_e\n",
      "EX_cytd_e\n",
      "EX_fru_e\n",
      "EX_man_e\n",
      "EX_rib__D_e\n",
      "EX_dad_2_e\n",
      "EX_dcyt_e\n",
      "EX_dgsn_e\n",
      "EX_etoh_e\n",
      "EX_fol_e\n",
      "EX_glyc_e\n",
      "EX_gly_e\n",
      "EX_gsn_e\n",
      "EX_h2o_e\n",
      "EX_hco3_e\n",
      "EX_ala__L_e\n",
      "EX_arg__L_e\n",
      "EX_asn__L_e\n",
      "EX_asp__L_e\n",
      "EX_cys__L_e\n",
      "EX_glu__L_e\n",
      "EX_gln__L_e\n",
      "EX_his__L_e\n",
      "EX_ile__L_e\n",
      "EX_leu__L_e\n",
      "EX_lys__L_e\n",
      "EX_met__L_e\n",
      "EX_orn_e\n",
      "EX_phe__L_e\n",
      "EX_pro__L_e\n",
      "EX_ser__L_e\n",
      "EX_thr__L_e\n",
      "EX_trp__L_e\n",
      "EX_tyr__L_e\n",
      "EX_val__L_e\n",
      "EX_acgam_e\n",
      "EX_na1_e\n",
      "EX_o2_e\n",
      "EX_pi_e\n",
      "SK_urdglyc_c\n",
      "EX_so4_e\n",
      "EX_thymd_e\n",
      "EX_urea_e\n",
      "EX_uri_e\n",
      "EX_gal_e\n",
      "EX_glc__D_e\n",
      "EX_xyl__D_e\n",
      "EX_h_e\n",
      "HMNO\n",
      "PGL\n",
      "GND\n",
      "AACOAT\n",
      "r0450\n",
      "ASPO1\n",
      "SACCD_c\n",
      "AASS_c\n",
      "ASPTA\n",
      "ASPTAm\n",
      "EHGLATm\n",
      "TYRTA\n",
      "ACACT10m\n",
      "ACCOAC\n",
      "ACACT1m\n",
      "ACOAH\n",
      "ACONTm\n",
      "ACONT\n",
      "ACS\n",
      "FACOAL40im\n",
      "MMCSm\n",
      "ASPCT\n",
      "ADA\n",
      "DADA\n",
      "ADNK1\n",
      "ADSS\n",
      "ALCD2x\n",
      "ALCD2y\n",
      "ADK1\n",
      "ADK1m\n",
      "DADK\n",
      "AGTim\n",
      "3AIBT2m\n",
      "AASAD3m\n",
      "ABUTD\n",
      "ALDD2x\n",
      "ALDD22x\n",
      "ALDH\n",
      "BETALDHxm\n",
      "XYLR\n",
      "ATAH_1\n",
      "ALLTN\n",
      "ALATA_L\n",
      "ALATA_Lm\n",
      "IZPN\n",
      "AGDC\n",
      "AMPD2\n",
      "GLYCLm\n",
      "GCCbim\n",
      "ADPT\n",
      "ARGN\n",
      "ARGNm\n",
      "ASNS1\n",
      "ACITL\n",
      "MGCHrm\n",
      "LEUTA\n",
      "VALTA\n",
      "ILETA\n",
      "ACOAD10m\n",
      "MOD_2hop\n",
      "MOD_pdhlam\n",
      "MOD_4mop\n",
      "MOD\n",
      "MOD_3mop\n",
      "MOD_3mhtpp\n",
      "MOD_2mhop\n",
      "MOD_2mbdhl\n",
      "GLUDC\n",
      "BUP\n",
      "HCO3E\n",
      "HCO3Em\n",
      "CYTD\n",
      "DCYTD\n",
      "CYSO\n",
      "CHOPT_PALM_PALM_c\n",
      "CHOLK\n",
      "CSm\n",
      "CKc\n",
      "CK\n",
      "CHLPCTD\n",
      "CPS_m\n",
      "CBPS\n",
      "HCGALm\n",
      "CAT\n",
      "CATm\n",
      "CTPS2\n",
      "CTPS1_2\n",
      "CYSTGL\n",
      "CYSTS\n",
      "DHLTAm\n",
      "DHRT_ivcoa\n",
      "DHRT_ibcoa\n",
      "MOD_dhlam\n",
      "ADMDC\n",
      "DCMPDA_1\n",
      "ADADir\n",
      "DRPA\n",
      "DHFR\n",
      "FOLR2_1\n",
      "DHORTS\n",
      "XYLTD_D\n",
      "PTRCOX1\n",
      "GCC2cm\n",
      "NTD6\n",
      "NTD8\n",
      "DURAD\n",
      "DURAD2\n",
      "DHPD\n",
      "DHPM2\n",
      "DTMPK\n",
      "URIDK2r\n",
      "HACD1m\n",
      "ECOAH1m\n",
      "ECOAH9m\n",
      "ECOAH12m\n",
      "HACD9m\n",
      "ENO\n",
      "ENOm\n",
      "NDP1\n",
      "NDP3\n",
      "ATPM\n",
      "NTP3\n",
      "NTP7\n",
      "NDP7\n",
      "NTP5\n",
      "CDPPH\n",
      "PSDOm\n",
      "FUMAC\n",
      "FAS\n",
      "FBA\n",
      "FBP\n",
      "GluForTx\n",
      "FTCD\n",
      "FTHFDH\n",
      "FTHFDHm\n",
      "FUMm\n",
      "FUM\n",
      "G6PP\n",
      "3A2OA\n",
      "ASP1DC\n",
      "GALKr\n",
      "UGLT\n",
      "GACMTRc\n",
      "GAPD\n",
      "GLYAMDTRc\n",
      "GLUTCOADHm\n",
      "GCCam\n",
      "GUAD\n",
      "GLUDy\n",
      "GDHm\n",
      "GLNS_1\n",
      "GALm\n",
      "GLUNm\n",
      "GLYK\n",
      "GLYCK2m\n",
      "G6PDA\n",
      "3SALATAi\n",
      "3SPYRSP\n",
      "G3PD1ir\n",
      "GLYCDH_1\n",
      "TRDR\n",
      "MACACI\n",
      "GMPS2\n",
      "GMPR\n",
      "GK1\n",
      "DGK1\n",
      "HISDr\n",
      "HGNTOR\n",
      "HIBH\n",
      "HIUH_c\n",
      "HMGLm\n",
      "HMGL\n",
      "HMGCOAS\n",
      "DDPGAm\n",
      "34HPPOR\n",
      "XPPT\n",
      "HXPRT\n",
      "GUAPRT\n",
      "HEX1\n",
      "HEX4\n",
      "ICDHxm\n",
      "ICDHym\n",
      "IMPD\n",
      "IDPh_1\n",
      "IDPm\n",
      "ACOAD8m\n",
      "CYSDSm\n",
      "OCOAT1m\n",
      "2OXOADOXm\n",
      "TMDK1\n",
      "DURIK1\n",
      "LDH_L\n",
      "LDH_Lm\n",
      "SUCOASm\n",
      "SUCOAS1m\n",
      "MCDm\n",
      "MCCCrm\n",
      "MMEm\n",
      "MDH\n",
      "MDHm\n",
      "ME2\n",
      "ME1m\n",
      "ME2m\n",
      "METAT\n",
      "MMSAD3\n",
      "MMSDH\n",
      "MAN6PI\n",
      "SK_5mta_c\n",
      "MTHFC\n",
      "MTHFCm\n",
      "MTHFD2m\n",
      "FTHFLi\n",
      "FTHFLmi\n",
      "MTHFD\n",
      "MTHFR2_1\n",
      "METS_1\n",
      "MMMm\n",
      "ACGAMK\n",
      "NDPK1\n",
      "NDPK8\n",
      "NDPK5\n",
      "NDPK2\n",
      "NDPK3\n",
      "NDPK7\n",
      "NDPK4\n",
      "NDPK9\n",
      "UMPK\n",
      "CYTK1\n",
      "CYTK2\n",
      "NTD11\n",
      "NTD7\n",
      "NTD10\n",
      "GMP5N\n",
      "NTD2\n",
      "NTD4\n",
      "ORNTArm\n",
      "ORNDC\n",
      "5HPUDICDC_c\n",
      "OCBTm\n",
      "HMR_4777\n",
      "PHCDm\n",
      "P5CRm\n",
      "G5SDym\n",
      "GLU5Km\n",
      "PHEH_c\n",
      "PPCOACm\n",
      "PEPCK_re\n",
      "PEPCKm\n",
      "ITPOXAL\n",
      "CRTNsyn\n",
      "PDHam1mi\n",
      "PDHam2mi\n",
      "PRFGS\n",
      "PFK\n",
      "PGK\n",
      "PGM\n",
      "PPM\n",
      "PPM2\n",
      "AGPATCOA_PALM_PALM_c\n",
      "G3PAT160_1\n",
      "PUNP1\n",
      "PUNP2\n",
      "PUNP6\n",
      "PUNP4\n",
      "DURIPP\n",
      "PAPA160\n",
      "r0074\n",
      "G5SADrm\n",
      "P5CDm\n",
      "PROD2m\n",
      "HPROxm\n",
      "PRPPS\n",
      "GLUPRT\n",
      "PRAGSr\n",
      "GARFT\n",
      "PRFGCL\n",
      "AIRCr\n",
      "PRASCSi\n",
      "ADSL2r\n",
      "ADSL1r\n",
      "AICART\n",
      "IMPC\n",
      "PC\n",
      "PCm\n",
      "PYK\n",
      "PYKm\n",
      "PYK5\n",
      "ORPT\n",
      "OMPDC\n",
      "NNOR\n",
      "RBK\n",
      "RNDR1\n",
      "RNDR2\n",
      "RNDR3\n",
      "RNDR4\n",
      "RPE\n",
      "RPI\n",
      "AHCi\n",
      "SARDHm\n",
      "ACOAD1fm\n",
      "ACOAD9m\n",
      "SUCD1m\n",
      "SERD_L\n",
      "PSERT\n",
      "PSP_L\n",
      "PGCD\n",
      "SGAT\n",
      "GHMT2r\n",
      "GHMT2rm\n",
      "SPMS\n",
      "SPRMS\n",
      "POLYAO3\n",
      "SQORm\n",
      "SSALxm\n",
      "r0142\n",
      "TALA\n",
      "THRA\n",
      "THRD_L\n",
      "TKT1\n",
      "TKT2\n",
      "TPI\n",
      "TMDS\n",
      "TMDPP\n",
      "PYNP2r\n",
      "UDPG4E\n",
      "ABTArm\n",
      "GALUi\n",
      "BUPN\n",
      "URO_c\n",
      "URIK1\n",
      "CYTDK1_1\n",
      "URCN\n",
      "PUNP5\n",
      "PUNP3\n",
      "PUNP7\n",
      "HXAND\n",
      "XYLK\n",
      "NaKt\n",
      "HKt\n",
      "ATPS4mi\n",
      "ORNt\n",
      "ATPS1\n",
      "2OBUTtm\n",
      "3AIBtm\n",
      "3MOBtm\n",
      "4MOPtm\n",
      "ATPtm_H\n",
      "ADNt4\n",
      "ASPGLUm\n",
      "GLCt2\n",
      "ALAt4\n",
      "ALAtmi\n",
      "ARGt2m\n",
      "ARGt4\n",
      "r2009\n",
      "r2008\n",
      "r2011\n",
      "r2015\n",
      "r2016\n",
      "ASNt4\n",
      "r1567\n",
      "r1553\n",
      "r1592\n",
      "r1630\n",
      "r1637\n",
      "ASPt6\n",
      "ASPt2m\n",
      "GLYBtm\n",
      "BUTt\n",
      "BUTt2m\n",
      "ACGAtex_1\n",
      "CHOLtu\n",
      "CHLtm\n",
      "MC_mal_pep\n",
      "HMR_4964\n",
      "CREATtmdiffir\n",
      "r0942\n",
      "CYSt4\n",
      "CYSt_m\n",
      "CYTDt4\n",
      "r1043\n",
      "DATPtn\n",
      "DCTPtn\n",
      "r1044\n",
      "DGTPtn\n",
      "r1045\n",
      "FUMtm\n",
      "MALtm\n",
      "DNAt\n",
      "DTTPtn\n",
      "ETOHt\n",
      "FORtm\n",
      "FRUt2\n",
      "4ABUTtm\n",
      "GALt2\n",
      "GLUt2m\n",
      "GTPt2m\n",
      "GLNtm_1\n",
      "GLNt4\n",
      "ALAGLNexR\n",
      "r1544\n",
      "SERGLNexR\n",
      "r1576\n",
      "CYSGLUexR\n",
      "GLUt6\n",
      "GLYt4\n",
      "GLYtm\n",
      "GLYCt\n",
      "G3PD2m\n",
      "GSNt4\n",
      "Ht\n",
      "H2Ot\n",
      "H2Otm\n",
      "HCO3t\n",
      "r0941\n",
      "HISt4\n",
      "ILEt4\n",
      "3MOPtm\n",
      "LEUt4\n",
      "LYSt4\n",
      "r1993\n",
      "r1992\n",
      "r1995\n",
      "r1999\n",
      "r2000\n",
      "MANt2\n",
      "L_LACtm\n",
      "PYRt2m\n",
      "METt4\n",
      "NH4t\n",
      "NH4tm\n",
      "O2t\n",
      "O2tm\n",
      "AKGMALtm\n",
      "ORNt4m\n",
      "PCREATtmdiffir\n",
      "PHEt4\n",
      "PIt2r\n",
      "PIt2m\n",
      "r1459\n",
      "PROt4\n",
      "r0911\n",
      "RIBt2\n",
      "SERt4\n",
      "r1435\n",
      "SO3t_m\n",
      "SO4ti\n",
      "SO4t_m\n",
      "THRt4\n",
      "THMDt4\n",
      "TRPt4\n",
      "TYRt4\n",
      "Htm\n",
      "UREAt\n",
      "UREAtm\n",
      "URIt4\n",
      "VALt4\n",
      "XYLt2\n",
      "ACACtm\n",
      "GLCS2\n",
      "EX_co2_e\n",
      "CO2t\n",
      "NAt\n",
      "CATp\n",
      "H2Otp\n",
      "H2O2tp\n",
      "O2tp\n",
      "Htx\n",
      "EX_nh4_e\n",
      "NH3c\n",
      "PGMT\n",
      "PGI\n",
      "G6PDH2r\n",
      "HEX7\n",
      "ACCOACrm\n",
      "CO2tm\n",
      "CHOLD2m\n",
      "CRTNtr\n",
      "EX_crtn_e\n",
      "GCC2bim\n",
      "GCCcm\n",
      "GCC2am\n",
      "PDHe2r\n",
      "OSUCCL\n",
      "ICITRED\n",
      "AKGDm\n",
      "THBPOX_m\n",
      "THBPt_m\n",
      "HTHBPt_m\n",
      "HTHBDH_m\n",
      "DHFOR\n",
      "FOLt\n",
      "DM_na1_c\n",
      "DM_fol_c\n",
      "CYOOm2i\n",
      "NADH2_u10mi\n",
      "CYOR_u10mi\n",
      "DHORD9\n",
      "NDPK6\n",
      "AATA\n",
      "2AMADPTm\n",
      "r2539\n",
      "2OXOADPTm\n",
      "ACYP\n",
      "PRAIS\n",
      "AMPDA\n",
      "ACGAMPM\n",
      "ACGAM2E\n",
      "UAG2EMA\n",
      "ACGAM6PS\n",
      "UAGDP\n",
      "GF6PTA\n",
      "CSNt\n",
      "EX_csn_e\n",
      "UPPN\n",
      "CSND\n",
      "DHPM1\n",
      "CITtbm\n",
      "XANDp\n",
      "HXANtx\n",
      "XAOx\n",
      "XAO2x\n",
      "XANtx\n",
      "URATEtx\n",
      "EX_no_e\n",
      "NOt\n",
      "NOS2\n",
      "NOS1\n",
      "ASNN\n",
      "FDH\n",
      "GNMT\n",
      "EX_sarcs_e\n",
      "SARCStex\n",
      "APAT2rm\n",
      "CARNS1\n",
      "BALAtmr\n",
      "MMSAD3m\n",
      "NBAHH_ir\n",
      "3SALAOX\n",
      "LCYSTCBOXL\n",
      "DM_taur_c\n",
      "EX_gthrd_e\n",
      "GTHRDt2\n",
      "GTHOr\n",
      "GTHS\n",
      "GLUCYS\n",
      "EX_cysi__L_e\n",
      "CYSGLTH\n",
      "CYSTSERex\n",
      "CYSTGLUex\n",
      "DM_hista_c\n",
      "HISDC\n",
      "OBDHc\n",
      "LEUTAm\n",
      "LEUt5m\n",
      "OIVD1m\n",
      "AADSACYCL\n",
      "LYStm\n",
      "SACCD3m\n",
      "AASAD3\n",
      "SACCD4m\n",
      "SARCStm\n",
      "ALDD19xr\n",
      "PHYCBOXL\n",
      "PEAMNO\n",
      "ALDD19x_P\n",
      "PACCOAL\n",
      "PHACCOAGLNAC\n",
      "EX_pheacgln_e\n",
      "PHEACGLNt\n",
      "THBPT4ACAMDASE\n",
      "PHETHPTOX2\n",
      "DHPR\n",
      "DHPR2\n",
      "NADtx\n",
      "NADHtpu\n",
      "NH4tp\n",
      "THP2Ctp\n",
      "1PPDCRp\n",
      "PPD2CSPp\n",
      "LPCOXp\n",
      "LYSOXp\n",
      "LYStip\n",
      "2MCITt\n",
      "EX_2mcit_e\n",
      "MCITS\n",
      "ANTHte\n",
      "EX_anth_e\n",
      "KYN\n",
      "TRPO2\n",
      "FKYNH\n",
      "KYNATESYN\n",
      "DM_kynate_c\n",
      "KYNAKGAT\n",
      "LFORKYNHYD\n",
      "r0239\n",
      "QULNtr\n",
      "EX_quln_e\n",
      "3HAO\n",
      "QUILSYN\n",
      "KYN3OX\n",
      "HKYNH\n",
      "DM_srtn_c\n",
      "5HLTDL\n",
      "TRPHYDRO2\n",
      "DHAAt1r\n",
      "DM_adrnl_c\n",
      "EX_ascb__L_e\n",
      "EX_dhdascb_e\n",
      "ASCBt\n",
      "3HLYTCL\n",
      "DOPABMO\n",
      "NORANMT\n",
      "TYRDOPO\n",
      "HIBDm\n",
      "VALt5m\n",
      "VALTAim\n",
      "MMTSADm\n",
      "OIVD2m\n",
      "3HBCOAHLm\n",
      "AM6SAD\n",
      "PCLAD\n",
      "AMCOXO\n",
      "ADSK\n",
      "BPNT2\n",
      "SADT\n",
      "DM_paps_c\n",
      "DUTPDP\n",
      "EX_ncam_e\n",
      "NMNAT\n",
      "NMNS\n",
      "NCAMUP\n",
      "SK_nad_c\n",
      "EX_ribflv_e\n",
      "SK_fad_c\n",
      "RIBFLVt3o\n",
      "RBFK\n",
      "RIBFLVt3\n",
      "FMNAT\n",
      "EX_pnto__R_e\n",
      "PNTOte\n",
      "SK_coa_c\n",
      "PNTK\n",
      "DPCOAK\n",
      "PPNCL3\n",
      "PPCDC\n",
      "PTPATi\n",
      "HMGCOASm\n",
      "PDHm\n",
      "SK_citr__L_c\n",
      "EX_glyb_e\n",
      "GLYBt4_2_r\n",
      "THYMt\n",
      "ADEt\n",
      "EX_ura_e\n",
      "EX_thym_e\n",
      "URAt\n",
      "EX_ade_e\n",
      "SPMDtex2\n",
      "EX_spmd_e\n",
      "FADH2ETC\n",
      "DNA_biomass\n",
      "RNA_biomass\n",
      "Lipids_biomass\n",
      "Protein_biomass\n",
      "Carbohydrates_biomass\n",
      "Biomass\n"
     ]
    }
   ],
   "source": [
    "import cobra\n",
    "from cobra.io import read_sbml_model\n",
    "\n",
    "model_path = \"C:\\\\Users\\\\salim\\\\OneDrive\\\\Documents\\\\AU\\\\System Biology & Mathematical Modeling\\\\Project\\\\salarecon.xml\"  # Your file path\n",
    "\n",
    "try:\n",
    "    model = read_sbml_model(model_path)\n",
    "    print(\"Model loaded successfully!\")\n",
    "\n",
    "    # Now you can work with the model object\n",
    "    # Example: print the model's reactions\n",
    "    print(\"Model reactions:\")\n",
    "    for reaction in model.reactions:\n",
    "       print(reaction.id)\n",
    "\n",
    "\n",
    "except cobra.exceptions.CobraSBMLError as e:\n",
    "    print(f\"Error loading SBML model: {e}\")\n",
    "except FileNotFoundError:\n",
    "    print(f\"Error: File not found at {model_path}\")\n",
    "except Exception as e:\n",
    "    print(f\"An unexpected error occurred: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00884371-e6ff-47d8-ac20-55978fffce8f",
   "metadata": {},
   "source": [
    "## Read Salmon gene table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed2015fd-81b0-423b-98ed-1eeb715fb370",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:4: SyntaxWarning: invalid escape sequence '\\S'\n",
      "<>:4: SyntaxWarning: invalid escape sequence '\\S'\n",
      "C:\\Users\\salim\\AppData\\Local\\Temp\\ipykernel_19948\\2683308610.py:4: SyntaxWarning: invalid escape sequence '\\S'\n",
      "  gene_table_path = \"C:\\\\Users\\\\salim\\\\OneDrive\\\\Documents\\\\AU\\\\System Biology & Mathematical Modeling\\\\Project\\Salmon_gene_table_NCBI.tsv\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Accession  Begin   End Chromosome Orientation  \\\n",
      "0  NC_001960.1   3846  4820         MT        plus   \n",
      "1  NC_001960.1   5036  6085         MT        plus   \n",
      "2  NC_001960.1   6476  8026         MT        plus   \n",
      "3  NC_001960.1   8190  8880         MT        plus   \n",
      "4  NC_001960.1   8956  9123         MT        plus   \n",
      "\n",
      "                              Name Symbol  Gene ID       Gene Type  \\\n",
      "0     NADH dehydrogenase subunit 1    ND1   808311  protein-coding   \n",
      "1     NADH dehydrogenase subunit 2    ND2   808316  protein-coding   \n",
      "2   cytochrome c oxidase subunit I   COX1   808314  protein-coding   \n",
      "3  cytochrome c oxidase subunit II   COX2   808315  protein-coding   \n",
      "4        ATP synthase F0 subunit 8   ATP8   808310  protein-coding   \n",
      "\n",
      "  Transcripts accession Protein accession  Protein length  Locus tag  \n",
      "0                   NaN       NP_008445.1             324        NaN  \n",
      "1                   NaN       NP_008446.1             349        NaN  \n",
      "2                   NaN       NP_008447.1             516        NaN  \n",
      "3                   NaN       NP_008448.1             230        NaN  \n",
      "4                   NaN       NP_008449.1              55        NaN  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Path to the gene table in TSV format\n",
    "gene_table_path = \"C:\\\\Users\\\\salim\\\\OneDrive\\\\Documents\\\\AU\\\\System Biology & Mathematical Modeling\\\\Project\\Salmon_gene_table_NCBI.tsv\"\n",
    "\n",
    "# Read the TSV file into a pandas DataFrame\n",
    "gene_table = pd.read_csv(gene_table_path, sep=\"\\t\")\n",
    "\n",
    "# Display the first few rows of the DataFrame to verify loading\n",
    "print(gene_table.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d54634f6-cb5b-45ca-a6a7-99927ec58d57",
   "metadata": {},
   "source": [
    "## Matching Gene IDs in Genome to get genes in the model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9127cbdb-685b-4375-b62f-023967cfa129",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            Accession  Begin    End Chromosome Orientation  \\\n",
      "0         NC_001960.1   3846   4820         MT        plus   \n",
      "1         NC_001960.1   5036   6085         MT        plus   \n",
      "7         NC_001960.1  10652  11000         MT        plus   \n",
      "9         NC_001960.1  11361  12741         MT        plus   \n",
      "10        NC_001960.1  12954  14792         MT        plus   \n",
      "...               ...    ...    ...        ...         ...   \n",
      "47964  NW_012378453.1    643   1856        NaN       minus   \n",
      "48187  NW_012387482.1    939   2610        NaN        plus   \n",
      "48286  NW_012393703.1    267   1983        NaN       minus   \n",
      "48357  NW_012405742.1     34   1815        NaN       minus   \n",
      "48487  NW_012435436.1    408    690        NaN        plus   \n",
      "\n",
      "                                                    Name        Symbol  \\\n",
      "0                           NADH dehydrogenase subunit 1           ND1   \n",
      "1                           NADH dehydrogenase subunit 2           ND2   \n",
      "7                           NADH dehydrogenase subunit 3           ND3   \n",
      "9                           NADH dehydrogenase subunit 4           ND4   \n",
      "10                          NADH dehydrogenase subunit 5           ND5   \n",
      "...                                                  ...           ...   \n",
      "47964  serine--pyruvate aminotransferase, mitochondri...  LOC106598450   \n",
      "48187                       thymidine phosphorylase-like  LOC106598843   \n",
      "48286                deoxyribose-phosphate aldolase-like  LOC106599083   \n",
      "48357                 mannose-6-phosphate isomerase-like  LOC106599274   \n",
      "48487           acyl carrier protein, mitochondrial-like  LOC106599655   \n",
      "\n",
      "         Gene ID       Gene Type Transcripts accession Protein accession  \\\n",
      "0         808311  protein-coding                   NaN       NP_008445.1   \n",
      "1         808316  protein-coding                   NaN       NP_008446.1   \n",
      "7         808308  protein-coding                   NaN       NP_008452.1   \n",
      "9         808305  protein-coding                   NaN       NP_008454.1   \n",
      "10        808309  protein-coding                   NaN       NP_008455.1   \n",
      "...          ...             ...                   ...               ...   \n",
      "47964  106598450  protein-coding        XM_014189491.1    XP_014044966.1   \n",
      "48187  106598843  protein-coding        XM_014189853.1    XP_014045328.1   \n",
      "48286  106599083  protein-coding        XM_014190139.1    XP_014045614.1   \n",
      "48357  106599274  protein-coding        XM_014190423.1    XP_014045898.1   \n",
      "48487  106599655  protein-coding        XM_014190941.1    XP_014046416.1   \n",
      "\n",
      "       Protein length  Locus tag  \n",
      "0                 324        NaN  \n",
      "1                 349        NaN  \n",
      "7                 116        NaN  \n",
      "9                 460        NaN  \n",
      "10                612        NaN  \n",
      "...               ...        ...  \n",
      "47964             180        NaN  \n",
      "48187             169        NaN  \n",
      "48286              83        NaN  \n",
      "48357             385        NaN  \n",
      "48487             157        NaN  \n",
      "\n",
      "[1136 rows x 13 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\salim\\AppData\\Local\\Temp\\ipykernel_19948\\1006700728.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  matched_rows[\"Symbol\"] = matched_rows[\"Symbol\"].str.upper()\n"
     ]
    }
   ],
   "source": [
    "gene_table[\"Gene ID\"] = gene_table[\"Gene ID\"].astype(str)\n",
    "# Example series of gene IDs to match\n",
    "gene_ids_to_match = [x.id for x in model.genes]  # Replace with your IDs\n",
    "\n",
    "# Extract rows where the 'GeneID' column matches the given IDs\n",
    "# Assuming the column containing gene IDs is named 'GeneID'\n",
    "matched_rows = gene_table[gene_table[\"Gene ID\"].isin(gene_ids_to_match)]\n",
    "matched_rows[\"Symbol\"] = matched_rows[\"Symbol\"].str.upper()\n",
    "print(matched_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "71cac4dd-00c3-4364-b9f0-c355a2bf0f9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        Seq Name  Number Reads                           Seq Description  \\\n",
      "0    contig02523          2787                        complement c4-like   \n",
      "1    contig04413          2740                                  ---NA---   \n",
      "2    contig01642          2693  vertebrate mid1 interacting-like protein   \n",
      "3    contig04278          2557                        complement c4-like   \n",
      "4    contig03717          2358                                  ---NA---   \n",
      "..           ...           ...                                       ...   \n",
      "749  contig01337            50                                  ---NA---   \n",
      "750  contig01510            50                           complement c4-1   \n",
      "751  contig01865            50                                  ---NA---   \n",
      "752  contig03659            50                 serum albumin 2 precursor   \n",
      "753  contig04133            50                 60s ribosomal protein l10   \n",
      "\n",
      "    Gene Name  Seq Length    Min eValue  Mean Similarity (%)  \n",
      "0         C4          435  9.590000e-60                 73.8  \n",
      "1         NaN         174           NaN                  NaN  \n",
      "2    MID1IP1          531  1.980000e-42                 63.2  \n",
      "3         C4          189  8.820000e-33                 75.8  \n",
      "4         NaN         272           NaN                  NaN  \n",
      "..        ...         ...           ...                  ...  \n",
      "749       NaN         574           NaN                  NaN  \n",
      "750      C4A          551  5.270000e-37                 77.4  \n",
      "751       NaN         506           NaN                  NaN  \n",
      "752       ALB         282  7.300000e-51                 66.9  \n",
      "753     RPL10         207  1.670000e-39                 97.4  \n",
      "\n",
      "[754 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "upregulated_gene_table = pd.read_csv(\"C:\\\\Users\\\\salim\\\\OneDrive\\\\Documents\\\\AU\\\\System Biology & Mathematical Modeling\\\\Project\\\\optimal_temp_13 - A) 19°C up-regulated.csv\", sep=\",\")\n",
    "print(upregulated_gene_table)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d8451f-ea17-449e-836a-1039222b41dc",
   "metadata": {},
   "source": [
    "## Matching Gene to Enzyme\n",
    "Analyze the gene expression data file to identify upregulated and downregulated genes.\n",
    "Determine which of these genes encode enzymes by cross-referencing gene annotations with enzyme present in the file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c59f2d72-402c-44e5-b53c-68ff600c7c5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Accession     Begin       End Chromosome Orientation  \\\n",
      "0  NC_027300.1  25099276  25112265      ssa01        plus   \n",
      "1  NC_027304.1  25596061  25599683      ssa05       minus   \n",
      "2  NC_027304.1  25596061  25599683      ssa05       minus   \n",
      "3  NC_027304.1  34712959  34717137      ssa05        plus   \n",
      "4  NC_027313.1  10963306  10974706      ssa14        plus   \n",
      "5  NC_027313.1  10963306  10974706      ssa14        plus   \n",
      "\n",
      "                                                Name   Symbol    Gene ID  \\\n",
      "0  methylmalonate-semialdehyde dehydrogenase [acy...  ALDH6A1  106592678   \n",
      "1              tryptophan 2,3-dioxygenase isoform X1     TDO2  106604524   \n",
      "2              tryptophan 2,3-dioxygenase isoform X1     TDO2  106604524   \n",
      "3     glyoxylate reductase/hydroxypyruvate reductase    GRHPR  100196601   \n",
      "4                  serine--pyruvate aminotransferase     AGXT  106568771   \n",
      "5                  serine--pyruvate aminotransferase     AGXT  106568771   \n",
      "\n",
      "        Gene Type Transcripts accession Protein accession  Protein length  \\\n",
      "0  protein-coding        XM_014184040.1    XP_014039515.1             526   \n",
      "1  protein-coding        XM_014199208.1    XP_014054683.1             407   \n",
      "2  protein-coding        XM_014199208.1    XP_014054683.1             407   \n",
      "3  protein-coding        NM_001141630.1    NP_001135102.1             328   \n",
      "4  protein-coding        XM_014139437.1    XP_013994912.1             428   \n",
      "5  protein-coding        XM_014139437.1    XP_013994912.1             428   \n",
      "\n",
      "   Locus tag     Seq Name  Number Reads  \\\n",
      "0        NaN  contig00638           184   \n",
      "1        NaN  contig04994            53   \n",
      "2        NaN  contig04995            51   \n",
      "3        NaN  contig03195           177   \n",
      "4        NaN  contig00248           117   \n",
      "5        NaN  contig00111            53   \n",
      "\n",
      "                                     Seq Description Gene Name  Seq Length  \\\n",
      "0          methylmalonate-semialdehyde dehydrogenase   ALDH6A1         683   \n",
      "1                            tryptophan -dioxygenase      TDO2         108   \n",
      "2                            tryptophan -dioxygenase      TDO2         108   \n",
      "3     glyoxylate reductase hydroxypyruvate reductase     GRHPR         348   \n",
      "4           serine--pyruvate mitochondrial precursor      AGXT         806   \n",
      "5  alanine--glyoxylate aminotransferase mitochond...      AGXT         932   \n",
      "\n",
      "      Min eValue  Mean Similarity (%)  \n",
      "0   3.150000e-69                 97.1  \n",
      "1   7.080000e-16                 92.2  \n",
      "2   7.080000e-16                 92.2  \n",
      "3   1.560000e-30                 88.8  \n",
      "4  2.460000e-164                 89.6  \n",
      "5   8.180000e-76                 86.1  \n"
     ]
    }
   ],
   "source": [
    "upregulated_gene_table = pd.read_csv(\"C:\\\\Users\\\\salim\\\\OneDrive\\\\Documents\\\\AU\\\\System Biology & Mathematical Modeling\\\\Project\\\\optimal_temp_13 - A) 19°C up-regulated.csv\")\n",
    "upregulated_gene_table[\"Gene Name\"] = upregulated_gene_table[\"Gene Name\"].str.upper()\n",
    "\n",
    "upregulated_metabolic_model = pd.merge(matched_rows, upregulated_gene_table, left_on=[\"Symbol\"], right_on=[\"Gene Name\"])\n",
    "print(upregulated_metabolic_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0b51ee5c-abfc-40ff-bce6-4d01f90b3ed7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Accession      Begin        End Chromosome Orientation  \\\n",
      "0  NC_027300.1  110478354  110565938      ssa01       minus   \n",
      "1  NC_027301.1   19308102   19348372      ssa02       minus   \n",
      "2  NC_027304.1   34712959   34717137      ssa05        plus   \n",
      "3  NC_027308.1  120445871  120460573      ssa09        plus   \n",
      "4  NC_027310.1   68685306   68686973      ssa11        plus   \n",
      "5  NC_027313.1   10963306   10974706      ssa14        plus   \n",
      "6  NC_027313.1   10963306   10974706      ssa14        plus   \n",
      "\n",
      "                                             Name Symbol    Gene ID  \\\n",
      "0          sarcosine dehydrogenase, mitochondrial  SARDH  106561570   \n",
      "1                             dihydropyrimidinase   DPYS  100380817   \n",
      "2  glyoxylate reductase/hydroxypyruvate reductase  GRHPR  100196601   \n",
      "3                   peroxisomal sarcosine oxidase  PIPOX  106613098   \n",
      "4               fatty acid-binding protein, liver  FABP1  106563252   \n",
      "5               serine--pyruvate aminotransferase   AGXT  106568771   \n",
      "6               serine--pyruvate aminotransferase   AGXT  106568771   \n",
      "\n",
      "        Gene Type Transcripts accession Protein accession  Protein length  \\\n",
      "0  protein-coding        XM_014125677.1    XP_013981152.1             929   \n",
      "1  protein-coding        NM_001173939.1    NP_001167410.1             362   \n",
      "2  protein-coding        NM_001141630.1    NP_001135102.1             328   \n",
      "3  protein-coding        XM_014215031.1    XP_014070506.1             386   \n",
      "4  protein-coding        XM_014128669.1    XP_013984144.1             126   \n",
      "5  protein-coding        XM_014139437.1    XP_013994912.1             428   \n",
      "6  protein-coding        XM_014139437.1    XP_013994912.1             428   \n",
      "\n",
      "   Locus tag     Seq Name  Number Reads  \\\n",
      "0        NaN  contig00066            56   \n",
      "1        NaN  contig00202            77   \n",
      "2        NaN  contig04120           120   \n",
      "3        NaN  contig00608           286   \n",
      "4        NaN  contig04644           272   \n",
      "5        NaN  contig00138           101   \n",
      "6        NaN  contig00406           101   \n",
      "\n",
      "                                  Seq Description Gene Name  Seq Length  \\\n",
      "0                    sarcosine mitochondrial-like     SARDH         866   \n",
      "1                             dihydropyrimidinase      DPYS         756   \n",
      "2  glyoxylate reductase hydroxypyruvate reductase     GRHPR         259   \n",
      "3                               loc558601 protein     PIPOX         646   \n",
      "4           liver-type fatty acid-binding protein     FABP1         188   \n",
      "5        serine--pyruvate mitochondrial precursor      AGXT         794   \n",
      "6             alanine-glyoxylate aminotransferase      AGXT         688   \n",
      "\n",
      "      Min eValue  Mean Similarity (%)  \n",
      "0   0.000000e+00                 85.9  \n",
      "1   1.070000e-83                 85.5  \n",
      "2   2.590000e-30                 87.1  \n",
      "3   1.500000e-15                 81.7  \n",
      "4   5.830000e-32                 95.5  \n",
      "5  7.370000e-165                 88.7  \n",
      "6  1.150000e-113                 85.2  \n"
     ]
    }
   ],
   "source": [
    "downregulated_gene_table = pd.read_csv(\"C:\\\\Users\\\\salim\\\\OneDrive\\\\Documents\\\\AU\\\\System Biology & Mathematical Modeling\\\\Project\\\\optimal_temp_13 - B) 19°C down-regulated.tsv\", sep=\"\\t\")\n",
    "downregulated_gene_table[\"Gene Name\"] = downregulated_gene_table[\"Gene Name\"].str.upper()\n",
    "\n",
    "downregulated_metabolic_model = pd.merge(matched_rows, downregulated_gene_table, left_on=[\"Symbol\"], right_on=[\"Gene Name\"])\n",
    "print(downregulated_metabolic_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "096b560f-c209-4099-81f6-fae20a5a34a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 1.613</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EX_adn_e</th>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-8.922042e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_but_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.021908e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_chol_e</th>\n",
       "      <td>-0.955420</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_creat_e</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EX_cytd_e</th>\n",
       "      <td>-1.000000</td>\n",
       "      <td>-4.248591e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA_biomass</th>\n",
       "      <td>0.011341</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lipids_biomass</th>\n",
       "      <td>0.701315</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein_biomass</th>\n",
       "      <td>0.806742</td>\n",
       "      <td>5.684342e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Carbohydrates_biomass</th>\n",
       "      <td>0.013883</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Biomass</th>\n",
       "      <td>1.613483</td>\n",
       "      <td>8.031770e-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>718 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 1.613 at 0x2aa7461f0b0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "solution = model.optimize()\n",
    "solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8bdbd31b-8dc1-48ab-b911-acb8b473b3cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene ID: 106561570\n",
      "Total Related Reactions: 1\n",
      "\n",
      "Detailed Reaction Fluxes:\n",
      "\n",
      "Reaction ID: SARDHm\n",
      "Reaction Name: Sarcosine dehydrogenase (m)\n",
      "Reaction equation: fad_m + sarcs_m + thf_m --> fadh2_m + gly_m + mlthf_m\n",
      "Flux Value: 0.0\n",
      "Lower Bound: 0.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 106561570\n",
      "\n",
      "-----------------------------------------------------------------------\n",
      "\n",
      "Gene ID: 100380817\n",
      "Total Related Reactions: 3\n",
      "\n",
      "Detailed Reaction Fluxes:\n",
      "\n",
      "Reaction ID: DHPD\n",
      "Reaction Name: Dihydropyrimidinase\n",
      "Reaction equation: 56dura_c + h2o_c <=> cala_c + h_c\n",
      "Flux Value: 0.0\n",
      "Lower Bound: -1000.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 100380817\n",
      "\n",
      "Reaction ID: DHPM2\n",
      "Reaction Name: Dihydropyrimidinase (dihydrothymine)\n",
      "Reaction equation: 56dthm_c + h2o_c <=> 3uib_c + h_c\n",
      "Flux Value: 0.0\n",
      "Lower Bound: -1000.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 100380817\n",
      "\n",
      "Reaction ID: DHPM1\n",
      "Reaction Name: Dihydropyrimidinase (5,6-dihydrouracil)\n",
      "Reaction equation: 56dura_c + h2o_c --> cala_c + h_c\n",
      "Flux Value: 0.0\n",
      "Lower Bound: 0.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 100380817 or 106567513 or 106565238 or 106563220 or 106579880 or 106585556\n",
      "\n",
      "-----------------------------------------------------------------------\n",
      "\n",
      "Gene ID: 100196601\n",
      "Total Related Reactions: 1\n",
      "\n",
      "Detailed Reaction Fluxes:\n",
      "\n",
      "Reaction ID: GLYCDH_1\n",
      "Reaction Name: Glycerate dehydrogenase, mitochondrial\n",
      "Reaction equation: h_m + hpyr_m + nadh_m <=> glyc__R_m + nad_m\n",
      "Flux Value: 0.0\n",
      "Lower Bound: -1000.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 100196601 or 106592840 or 106611699\n",
      "\n",
      "-----------------------------------------------------------------------\n",
      "\n",
      "Gene ID: 106613098\n",
      "Total Related Reactions: 1\n",
      "\n",
      "Detailed Reaction Fluxes:\n",
      "\n",
      "Reaction ID: LPCOXp\n",
      "Reaction Name: L-pipecolate oxidase, peroxisomal\n",
      "Reaction equation: Lpipecol_x + o2_x --> h2o2_x + h_x + thp2c_x\n",
      "Flux Value: 0.0\n",
      "Lower Bound: 0.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 106613098\n",
      "\n",
      "-----------------------------------------------------------------------\n",
      "\n",
      "Gene ID: 106563252\n",
      "Total Related Reactions: 1\n",
      "\n",
      "Detailed Reaction Fluxes:\n",
      "\n",
      "Reaction ID: BUTt\n",
      "Reaction Name: Butyrate transport (diffusion)\n",
      "Reaction equation: but_e --> but_c\n",
      "Flux Value: 0.0\n",
      "Lower Bound: 0.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 106571818 or 106570598 or 100136444 or 106601642 or 106606772 or 106609495 or 106576427 or 106571805 or 106608508 or 106602733 or 106610011 or 106563252 or 100195728 or 100196877 or 100862707 or 106588813 or 106570571 or 100196174 or 106593471 or 100196451 or 106561537 or 106568426 or 106603912 or 106611950 or 106567475 or 100196451 or 106588813 or 106570571 or 100196174 or 106593471 or 100196451 or 106588813 or 106570571 or 100196174 or 106593471 or 100196451\n",
      "\n",
      "-----------------------------------------------------------------------\n",
      "\n",
      "Gene ID: 106568771\n",
      "Total Related Reactions: 1\n",
      "\n",
      "Detailed Reaction Fluxes:\n",
      "\n",
      "Reaction ID: SGAT\n",
      "Reaction Name: Serine--glyoxylate aminotransferase\n",
      "Reaction equation: glx_m + ser__L_m <=> gly_m + hpyr_m\n",
      "Flux Value: 0.0\n",
      "Lower Bound: -1000.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 106568771 or 106574201 or 106598450\n",
      "\n",
      "-----------------------------------------------------------------------\n",
      "\n",
      "Gene ID: 106568771\n",
      "Total Related Reactions: 1\n",
      "\n",
      "Detailed Reaction Fluxes:\n",
      "\n",
      "Reaction ID: SGAT\n",
      "Reaction Name: Serine--glyoxylate aminotransferase\n",
      "Reaction equation: glx_m + ser__L_m <=> gly_m + hpyr_m\n",
      "Flux Value: 0.0\n",
      "Lower Bound: -1000.0\n",
      "Upper Bound: 1000.0\n",
      "Gene Reaction Rule: 106568771 or 106574201 or 106598450\n",
      "\n",
      "-----------------------------------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Store reactions and their fluxes\n",
    "def get_gene_reaction_fluxes(solution, gene_id):\n",
    "    gene_related_reactions = []\n",
    "\n",
    "    # Iterate through all reactions in the model\n",
    "    for reaction in model.reactions:\n",
    "        # Check if the gene is associated with this reaction\n",
    "        if any(str(gene_id) == str(gene.id) for gene in reaction.genes):\n",
    "            # Get the flux value for this reaction\n",
    "            flux_value = solution.fluxes.get(reaction.id, 0)\n",
    "\n",
    "            gene_related_reactions.append({\n",
    "                'reaction_id': reaction.id,\n",
    "                'reaction_name': reaction.name,\n",
    "                'reaction equation':reaction.reaction,\n",
    "                'flux_value': flux_value,\n",
    "                'lower_bound': reaction.lower_bound,\n",
    "                'upper_bound': reaction.upper_bound,\n",
    "                'gene_reaction_rule': reaction.gene_reaction_rule\n",
    "            })\n",
    "\n",
    "    return {\n",
    "        'gene_id': gene_id,\n",
    "        'total_related_reactions': len(gene_related_reactions),\n",
    "        'reactions': gene_related_reactions\n",
    "    }\n",
    "\n",
    "# Example usage\n",
    "# Assuming 'solution' is your already loaded metabolic solution\n",
    "gene_ids = [\"106561570\", \"100380817\", \"100196601\", \"106613098\", \"106563252\", \"106568771\", \"106568771\"]  # Replace with your specific gene ID\n",
    "for gene_id in gene_ids:\n",
    "    result = get_gene_reaction_fluxes(solution, gene_id)\n",
    "\n",
    "    # Print detailed results\n",
    "    print(f\"Gene ID: {result['gene_id']}\")\n",
    "    print(f\"Total Related Reactions: {result['total_related_reactions']}\")\n",
    "    print(\"\\nDetailed Reaction Fluxes:\")\n",
    "    for reaction in result['reactions']:\n",
    "        print(\"\\nReaction ID:\", reaction['reaction_id'])\n",
    "        print(\"Reaction Name:\", reaction['reaction_name'])\n",
    "        print(\"Reaction equation:\", reaction['reaction equation'])\n",
    "        print(\"Flux Value:\", reaction['flux_value'])\n",
    "        print(\"Lower Bound:\", reaction['lower_bound'])\n",
    "        print(\"Upper Bound:\", reaction['upper_bound'])\n",
    "        print(\"Gene Reaction Rule:\", reaction['gene_reaction_rule'])\n",
    "        \n",
    "    print(\"\\n-----------------------------------------------------------------------\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a40e1c8-ed2c-43bb-b91b-8afe36920f2a",
   "metadata": {},
   "source": [
    "### Change upperbond for SARDHm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2bd15736-7a33-44ff-bced-e53cdc2a7089",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"SARDHm\").upper_bound = 500\n",
    "solution_SARDHm_over = model.optimize()\n",
    "solution_SARDHm_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db661758-4f16-4afa-90f6-13db44e6224f",
   "metadata": {},
   "source": [
    "### Change upperbond for DHPD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9b2f6937-d54c-4f30-b4e2-ace7ae2f2875",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"DHPD\").upper_bound = 500\n",
    "solution_DHPD_over = model.optimize()\n",
    "solution_DHPD_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d01e5975-cf89-4dc1-a9db-4c28f4772ba4",
   "metadata": {},
   "source": [
    "### Change upperbond for DHPM2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e03ba593-43a2-423a-a896-a210c3bb4e63",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"DHPM2\").upper_bound = 500\n",
    "solution_DHPM2_over = model.optimize()\n",
    "solution_DHPM2_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbfc251a-e654-4af7-a3a2-5331bdd31757",
   "metadata": {},
   "source": [
    "### Change upperbond for DHPM1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "20d91742-af7f-416a-8e91-a9085d5d5e1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"DHPM1\").upper_bound = 500\n",
    "solution_DHPM1_over = model.optimize()\n",
    "solution_DHPM1_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb868d8f-5417-4b0c-ade3-ee1e5c80e544",
   "metadata": {},
   "source": [
    "### Change upperbond for GLYCDH_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4ae45b1a-0e0e-41d0-9d32-37ca5c0c5df2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"GLYCDH_1\").upper_bound = 500\n",
    "solution_GLYCDH_1_over = model.optimize()\n",
    "solution_GLYCDH_1_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67eb2c14-b851-4a9a-b606-826763fd8a52",
   "metadata": {},
   "source": [
    "### Change in LPCOXp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "95e4695e-98ec-4fff-83de-c4a9e6ea1904",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"LPCOXp\").upper_bound = 500\n",
    "solution_LPCOXp_over = model.optimize()\n",
    "solution_LPCOXp_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a8b8d94-f5ec-4cde-9aef-046449f9c41b",
   "metadata": {},
   "source": [
    "### Change in BUTt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "87fb6417-b7a7-44e5-8630-ff9faf7d4ba1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"BUTt\").upper_bound = 500\n",
    "solution_BUTt_over = model.optimize()\n",
    "solution_BUTt_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "837ff618-bb45-4898-b3b4-13b1282b0029",
   "metadata": {},
   "source": [
    "### Change in SGAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "50f19f1c-beba-4bbc-99a8-3a836e96c460",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.float64(1.6134830016392714)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id(\"SGAT\").upper_bound = 500\n",
    "solution_SGAT_over = model.optimize()\n",
    "solution_SGAT_over[\"Biomass\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b90c3515-bde2-4493-b99f-6362de25a60d",
   "metadata": {},
   "source": [
    "### Conclusion\n",
    "No change was observed when changing the upper bond from 1000 to 500 for all the genes\n",
    "\n",
    "The fluxes for down-regulated enzyme's reactions is  0 in the original steady state, then we can say that probably the model is not 100% accurate."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
